src/Features/LanguageServer/Protocol/Handler/CodeActions/CodeActionsHandler.cs

```diff
@@ -48,56 +49,76 @@ public CodeActionsHandler(ICodeFixService codeFixService, ICodeRefactoringServic
             var result = new ArrayBuilder<object>();
             foreach (var codeAction in codeActions)
             {
+                if (!clientSupportsWorkspaceEdits)
+                {
+                    AddCodeActionWithoutIncludingChanges(request, result, codeAction);
+                    continue;
+                }
+
                 // If we have a codeaction with a single applychangesoperation, we want to send the codeaction with the edits.
                 var operations = await codeAction.GetOperationsAsync(cancellationToken).ConfigureAwait(keepThreadContext);
```

Comment:
Might be good to add a comment explaining that this line triggers the majority of the work, so only do it if necessary.

Reply:
This line is gone now. :)